Add back the link to the deployed App

xoen opened the pull request:
### What
This used to be part of the frontend but got lost in translation when removed node app.

This is adding it back:

[Imagine gif of Bender saying "I'm back, baby" here]

### How
The URL will be the host in the `Ingress` with the `repo` label set to the app repo (repo name without the URL).

As the `Ingress` resources don't live in the users' namespaces, users will not have sufficient permissions to list them, therefore this request is performed using the CP credentials (which can list Ingresses in that namespace). 
In order to do that I've tweaked the `api.kubernetes.KubernetesClient` class to receive a `use_cpanel_creds` argument (default is `False` as use of CP creds is discouraged and we should always aim at making requests to the k8s API using the user ID token).

### Small refactorings, etc...
- following conversation with @andyhd I've started to group the app-related operations performed on the "cluster" (k8s cluster but also AWS account) into the `api.cluster.App` class which wraps the `api.models.App`. Future plans include splitting this big module into smaller files (maybe rename it `external` to avoid further confusion with the k8s cluster?)
- `api.kubernetes.KubernetesClient` class now blows up if:
  - `id_token` is not provided (it optional) but `user_cpanel_creds=False` as this would cause the  up using the client to use the host (in-cluster or kube config) credentials
  - both `id_token` and `user_cpanel_creds=True` are provided as this is ambiguous

### Future refactorings
In the spirit of trying to keep PRs small and as focused as possible (and following @josh always wise advice), I've refrained myself from further simplifying the `api.kubernetes.KubernetesClient` class (by potentially removing logic to read session `id_token` and legacy RBAC flag) as it turned out to be more involved than expected - so I may do that separately

### Ticket
https://trello.com/c/eQqJ1FRw

andyhd commented on controlpanel/api/cluster.py:
```diff
@@ -124,17 +124,45 @@ def purge_user(user):
+
+    @property
+    def iam_role_name(self):
+        return f"{settings.ENV}_app_{self.app.slug}"
```
👍

andyhd commented on controlpanel/api/cluster.py:
```diff
@@ -124,17 +124,45 @@ def purge_user(user):
+        ).items
+
+        if len(ingresses) != 1:
+            return ""
```
Returning an empty string is fine as it evaluates to `False` in `{% if app_url %}`, but I think it makes more sense to return `None`

andyhd replied:
Also, if there are more than one ingress, shouldn't we return one of them? It will be a valid URL, won't it?

xoen replied:
I'll have a think about the `None` but it may be better as `None` conveys better the "no URL" meaning.

xoen replied:
Regarding the multiple ingress I'd stick to return `""`/`None` as there should be only a single `Ingress` per app so if two are found we can't be 100% sure which URL is the right one.

I'd rather have the user come to us and ask why the URL is not shown and we then investigate than have shown a potentially wrong URL, the user freaks out, it comes to us anyway and we freak out as well.

It shouldn't happen anyway but I prefer to not pick a "random" URL when there is potential ambiguity.

andyhd replied:
In my opinion, a working but "wrong" URL would be better than a missing one, but OK

andyhd commented on controlpanel/api/kubernetes.py:
```diff
@@ -9,10 +9,30 @@
+                "Remember: avoid using the Control Panel credentials to talk with "
+                "the k8s API unless stricly necessary."
+            )
+
```
I think you've forgotten to push some code here

xoen replied:
Oh god, you're absolutely right! Superstar for catching this @andyhd 🌟❤️ 

I was tweaking this class and eventually decided to move the refactoring experiment into a separate PR but I thrown away the feature with the dirty water 🤦‍♂ 

The missing code is simply a `and not use_cpanel_creds` added to the `if`:

```python
        if settings.ENABLED['k8s_rbac'] and not use_cpanel_creds:
            if id_token is None:
                # ...
```

as in that case we don't want to use the `id_token`.

I'll dig this out or add again. Amazing, thanks again for spotting this Andy ❤️

andyhd commented on controlpanel/api/models/app.py:
```diff
@@ -72,10 +72,10 @@ def save(self, *args, **kwargs):
 
         if is_create:
-            cluster.create_app_role(self.iam_role_name)
+            cluster.App(self).create_iam_role()
```
I like this 👍

andyhd commented on doc/running.md:
```diff
@@ -113,3 +113,13 @@ Go to http://localhost:8000/
+
+**NOTE** Set the `DJANGO_SETTINGS_MODULE` is important or otherwise you
+may accidentally run the tests with the `development` settings with
+unpredictable results.
```
The `DJANGO_SETTINGS_MODULE` used by `pytest` is configured in `/setup.cfg`, you shouldn't have to set it manually - is that not working?

andyhd replied:
Also, you don't need to pass `--color=yes`, colours are automatically enabled if the terminal is capable of showing them

xoen replied:
When I got this problem I found out that yes this is configured in `./setup.cfg` but in my case I also had that `DJANGO_SETTINGS_MODULE` set in my environment variable which unfortunately (in this case!) takes precedence against the `setup.cfg`.

This is by design and makes a lot of sense but I imagined it could happen to others following this documentation as one of the first things at the top of the page is to set `DJANGO_SETTINGS_MODULE` to `*.development`. I just thought having a simple note here could save someone else some time in the future.

xoen replied:
Regarding `--color=yes` I'll remove, I just copied whatever was in the `Makefile`.

andyhd commented on tests/api/cluster/test_app.py:
```diff
@@ -0,0 +1,74 @@
+    k8s_client \
+        .ExtensionsV1beta1Api \
+        .list_namespaced_ingress \
+        .assert_called_once_with("apps-prod", label_selector="repo=test-repo")
```
I'm pretty sure you can and should remove these `\`s - but use [Black](https://black.readthedocs.io/en/stable/index.html) as the source of truth on formatting

andyhd replied:
You can break the line after `(` like:
```python
    k8s_client.ExtensionsV1beta1Api.list_namespaced_ingress.assert_called_once_with(
        "apps-prod",
        label_selector="repo=test-repo",
    )
```
The long line is <88 chars, which is the default max line-length in Black.

xoen replied:
@r4vi (who was there when I was writing this) told me to tell you I run this through black :trollface: 

I'll play with it. (no wonder Go's autoformatting is everyone's favourite feature 😎)

andyhd commented on tests/api/cluster/test_app.py:
```diff
@@ -0,0 +1,74 @@
+    k8s_client \
+        .ExtensionsV1beta1Api \
+        .list_namespaced_ingress \
+        .assert_called_once_with("apps-prod", label_selector="repo=test-repo")
```
You should replace these last three tests with a single parameterised test. Eg:
```python
mock_ingress = MagicMock(name="Ingress")
mock_ingress.spec.rules = [MagicMock(name="Rule", host="test-app.example.com")]

@pytest.mark.parameterize(
    'ingresses, expected_url',
    [
        ([], "",),
        (["ing_1", "ing_2"], ""),
        ([mock_ingress], "https://test-app.example.com"),
    ],
    ids=[
        "no-ingresses",
        "multiple-ingresses",
        "single-ingress",
    ]
)
def test_app_url(k8s_client, app, app_model, ingresses, expected_url):
    list_ingresses = k8s_client.Extensions.V1beta1Api.list_namespaced_ingress
    list_ingresses.return_value.items = ingresses
    assert app_url == expected_url
    list_ingresses.assert_called_once_with("apps-prod", label_selector="repo=test-repo")
```

xoen replied:
Nice. I noticed at the time that there was some repetition in these tests but I got frustrated at pytest and when it worked and it was nice enough (you should have seen the previous 2 drafts of this test 😆) I left it.

Thanks for the hint ❤️ , I'll play with it 👍

xoen replied:
```pytest
test_app_url has 'parameterize' mark, spelling should be 'parametrize'
```

blimey, "parametrize" that's a difficult word 🤣

xoen commented on controlpanel/api/kubernetes.py:
```diff
@@ -9,15 +9,34 @@
 
-        if settings.ENABLED['k8s_rbac']:
-
+        if settings.ENABLED["k8s_rbac"] and not use_cpanel_creds:
```
@andyhd this is the bit that somehow got lost in previous commits set.

xoen commented:
@andyhd have another look if you wish, I've rebased and added back the missing code using the `use_cpanel_creds` argument.

Also:
- `cluster.App.url` returns `None` when can't determine URL
- refactored the tests to use parametrize 🏆 
- and formatted some of the code with black
